<a href="https://colab.research.google.com/github/Al1veeee/Notebook-python-/blob/main/Alimkhanov_D_Workbook2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1> РАБОЧАЯ ТЕТРАДЬ 2</h1>
<h2> Алимханов Дени ИКБО-68-23</h2>
<h1>1.1. Теоретический материал – Библиотека NumPy</h1>
<p>NumPy (NumericalPython) - это библиотека Python с открытым
исходным кодом, которая используется практически во всех областях науки
и техники. Это универсальный стандарт для работы с числовыми данными в
Python. </br>
Если у вас уже есть Python, вы можете установить NumPy с помощью
командной строки:</br>
 pip install numpy</br>
Чтобы начать использовать NumPy необходимо импортировать
соответствующую библиотеку:</br>
 import numpy as np</br>
Основным объектом NumPy является однородный многомерный
массив (в numpy называется numpy.ndarray). Это многомерный массив
элементов (обычно чисел), одного типа.</br>
Наиболее важные атрибуты объектов ndarray:</br>
ndarray.ndim - число измерений (чаще их называют "оси") массива.</br>
ndarray.shape - размеры массива, его форма. Это кортеж натуральных
чисел, показывающий длину массива по каждой оси. Для матрицы из n строк
и m столбов, shape будет (n,m). Число элементов кортежа shape равно ndim.</br>
ndarray.size - количество элементов массива. Очевидно, равно
произведению всех элементов атрибута shape.</br>
ndarray.dtype - объект, описывающий тип элементов массива. Можно
определить dtype, используя стандартные типы данных Python. NumPy здесь
предоставляет целый букет возможностей, как встроенных, например: bool_,
character, int8, int16, int32, int64, float8, float16, float32, float64, complex64,
object_, так и возможность определить собственные типы данных, в том
числе и составные.</br>
ndarray.itemsize - размер каждого элемента массива в байтах.</br>
ndarray.data - буфер, содержащий фактические элементы массива.</br>
Обычно не нужно использовать этот атрибут, так как обращаться к
элементам массива проще всего с помощью индексов.</br>
Подробнее о массивах в NumPy можно найти в официальной
документации https://numpy.org/doc/stable/user/absolute_beginners.html
</p>
<h1>1.2.1 Пример </h1>
<p>Создать массив 5x2. Создать массив 5x2. Вывести все значения массива,
значение элемента с индексом (3,1) и второй столбец. Индексация
начинается с нуля.</p>

In [ ]:
import numpy as np
x = np.array([[1,2],[3,4],[5,6],[7,8],[9,10]])
print (x)
print (x[3][1])
print (x[1])

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]]
8
[3 4]


<h1> 1.2.2 Пример </h1>
<p>Пример. Выполнить следующее: </br>
1. Создать вектор (одномерный массив) размера 10, заполненный
нулями.</br>
2. Создать вектор размера 10, заполненный единицами.</br>
3. Создать вектор размера 10, заполненный заданным числом.</br>
4. Создать вектор со значениями от 10 до 19.</p>

In [ ]:
import numpy as np
a = np.zeros(10)
b = np.ones(10)
c = np.full(10,5)
d = np.arange(10,20)
print(a, "\n", b, "\n", c, "\n", d)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.] 
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.] 
 [5 5 5 5 5 5 5 5 5 5] 
 [10 11 12 13 14 15 16 17 18 19]


<h1>1.2.3 Пример</h1>
<p>Создать массив 10x10 со случайными значениями, найти минимум, максимум и среднее значение.</p>

In [ ]:
import numpy as np
Z = np.random.random((10,10))
Zmin, Zmax, Zmean = Z.min(),Z.max(),Z.mean()
print(Zmin, Zmax, Zmean)

0.030348213048374206 0.9860381867647092 0.506646662425408


<h1>1.2.4 Пример </h1>
<p>Задать матрицу размерности 5 на 5 и поменять 2 строки в матрице местами.</p>

In [ ]:
import numpy as np
A = np.arange(25).reshape(5,5)
A[[0,1]] = A[[1,0]]
print(A)

[[ 5  6  7  8  9]
 [ 0  1  2  3  4]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


<h1>1.2.5 Пример </h1>
<p>Выяснить результат следующих выражений: </br>
0 * np.nan</br>
np.nan == np.nan</br>
np.inf > np.nan</br>
np.nan - np.nan</br>
0.3 == 3 * 0.1</p>

In [ ]:
import numpy as np
print(0 * np.nan)
print(np.nan == np.nan)
print(np.inf > np.nan)
print(np.nan - np.nan)
print(0.3 == 3*0.1)

nan
False
False
nan
False


<h1>1.2.6 Пример </h1>
<p>Отсортировать массив</p>

In [ ]:
import numpy as np
arr = np.array ([2,1,5,3,7,4,6,8])
print(np.sort(arr))

[1 2 3 4 5 6 7 8]


<h1>1.3.1 Задание</h1>
<p>Создать 8x8 матрицу и заполнить её в шахматном порядке нулями и единицами.</p>

In [ ]:
import numpy as np
A = np.zeros((8,8), dtype=int)
A[1::2,::2] = 1
A[::2,1::2] = 1
print(A)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


<h1>1.3.2 Задание</h1>
<p>Создать 5x5 матрицу со значениями в строках от 0 до 4.</p>

In [ ]:
import numpy as np
A = np.zeros((5,5), dtype=int)
A += np.arange(5)
print(A)

[[0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]]


<h1>1.3.3 Задание</h1>
<p>Создать массив 3x3x3 со случайными значениями.</p>

In [ ]:
import numpy as np
Z = np.random.random((3,3,3))
print(Z)

[[[0.22276218 0.30523741 0.27514458]
  [0.42997277 0.9923198  0.0425786 ]
  [0.29785735 0.31664267 0.50703285]]

 [[0.23370509 0.92949103 0.19411883]
  [0.09502712 0.28391101 0.92530912]
  [0.93689671 0.05929806 0.67841251]]

 [[0.80921487 0.79950413 0.29880361]
  [0.9760213  0.02157457 0.91196447]
  [0.56253925 0.50409317 0.61210408]]]


<h1>1.3.4 Задание</h1>
<p>Создать матрицу с 0 внутри, и 1 на границах.</p>

In [ ]:
import numpy as np
A = np.ones((5,5), dtype=int)
A[1:-1,1:-1] = 0
print(A)

[[1 1 1 1 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 1 1 1 1]]


<h1>1.3.5 Задание</h1>
<p>Создайте массив и отсортируйте его по убыванию.</p>

In [ ]:
A = [3.4, 5.1, 2.2, 4.1, 1.0, 3.8]

A.sort(reverse=True)
print(A)

[5.1, 4.1, 3.8, 3.4, 2.2, 1.0]


<h1>1.3.6 Задание</h1>
<p>Создайте матрицу, выведите ее форму, размер и размерность.</p>

In [ ]:
import numpy as np
size = int(input("Размерность - "))
A = np.random.random((size,size))
print("Форма -",A.shape)
print("Размер -",A.size)
print("Размерность -",A.ndim)

Размерность - 5
Форма - (5, 5)
Размер - 25
Размерность - 2


<h1>2.1. Теоретический материал – Библиотека Pandas</h1>
<p>Первым шагом в любом начинании в области машинного обучения
является введение исходных данных в систему. Исходные данные могут
вводиться вручную, содержаться в файле или храниться в интернете в какомлибо формате. Кроме того, часто требуется получить данные из нескольких
источников. </br>
Библиотека pandas – это удобный и быстрый инструмент для работы
с данными, обладающий большим функционалом. Если очень кратко, то
pandas – это библиотека, которая предоставляет очень удобные с точки
зрения использования инструменты для хранения данных и работе с ними.
Библиотека pandas присутствует в стандартной поставке Anaconda.
Если же ее там нет, то его можно установить отдельно. Для этого введите
командной строке: </br>
 pip install pandas</br>
Для импорта библиотеки используйте команду:</br>
import pandas as pd</br>
Библиотека pandas предоставляет две ключевые структуры данных:</br>
Series и DataFrame.</br>
Series – это одномерная структура данных, ее можно представить, как
таблицу с одной строкой. С Series можно работать как с обычным массивом
(обращаться по номеру индекса), и как с ассоциированным массивом, когда
можно использовать ключ для доступа к элементам данных.</br>
DataFrame – это двумерная структура. Идейно она очень похожа на
обычную таблицу, что выражается в способе ее создания и работе с ее
элементами</p></br>
<h1>2.2.1 Пример</h1>
<p>Создать Series из списка Python, словаря Python, и массива Numpy (установить буквенные метки для последнего). </p>

In [ ]:
import pandas as pd
lst = [1, 2, 3, 4, 5]
d = {'a':1, 'b':2, 'c':3}
ndarr = np.array([1, 2, 3, 4, 5])

s1 = pd.Series(lst)
s2 = pd.Series(d)
s3 = pd.Series(ndarr, ['a', 'b', 'c', 'd', 'e'])

print(s1)
print(s2)
print(s3)

0    1
1    2
2    3
3    4
4    5
dtype: int64
a    1
b    2
c    3
dtype: int64
a    1
b    2
c    3
d    4
e    5
dtype: int32


<h1>2.2.2 Пример</h1>
<p>Дано два Series. Напечатать их первые элементы и все элементы после третьего (во втором фрейме).</p>

In [ ]:
import pandas as pd
s1 = pd.Series([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e'])
s2 = pd.Series([5, 4, 3, 2, 1])
print(s1['a'])
print(s2[0])
print(s2[3:])

1
5
3    2
4    1
dtype: int64


<h1>2.2.3 Пример </h1>
<p>Создайте новый фрейм данных.</p>

In [ ]:
import pandas as pd
dataframe = pd.DataFrame()
dataframe['Имя'] = ['Джеки Джексон', 'Стивен Стивенсон']
dataframe['Возраст'] = [38, 25]
dataframe['Водитель'] = [True, False]
dataframe

Имя  Возраст  Водитель
0     Джеки Джексон       38      True
1  Стивен Стивенсон       25     False

<h1>2.2.4 Пример</h1>
<p>Загрузите фрейм данных по ссылке:
https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titan
ic.csv</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
dataframe = pd.read_csv(url) #Загрузить данные
dataframe.head(5) #Показать пять строк

Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92    male   

   Survived  SexCode  
0         1        1  
1         0        1  
2         0        0  
3         0        1  
4         1        0

<h1> 2.2.5 Пример </h1>
<p>Пронализировать характеристики фрейма данных.</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
dataframe = pd.read_csv(url)
display(dataframe.head(5))
display(dataframe.tail(3))
display(dataframe.shape)
display(dataframe.describe())

Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92    male   

   Survived  SexCode  
0         1        1  
1         0        1  
2         0        0  
3         0        1  
4         1        0

Name PClass   Age   Sex  Survived  SexCode
1310  Zenni, Mr Philip    3rd  22.0  male         0        0
1311  Lievens, Mr Rene    3rd  24.0  male         0        0
1312    Zimmerman, Leo    3rd  29.0  male         0        0

(1313, 6)

Age     Survived      SexCode
count  756.000000  1313.000000  1313.000000
mean    30.397989     0.342727     0.351866
std     14.259049     0.474802     0.477734
min      0.170000     0.000000     0.000000
25%     21.000000     0.000000     0.000000
50%     28.000000     0.000000     0.000000
75%     39.000000     1.000000     1.000000
max     71.000000     1.000000     1.000000

<h1>2.2.6 Пример</h1>
<p>Выберите индивидуальные данные или срезы фрейма данных.</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
dataframe = pd.read_csv(url)
dataframe.iloc[1:4]

Name PClass   Age     Sex  \
1                    Allison, Miss Helen Loraine    1st   2.0  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.0    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.0  female   

   Survived  SexCode  
1         0        1  
2         0        0  
3         0        1

<h1> 2.2.7 Пример </h1>
<p>Требуется отобрать строки фрейма данных на основе некоторого
условия. Необходимо сформировать новый фрейм данных из пассажиров
первого класса.</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'
dataframe = pd.read_csv(url)
dataframe[dataframe['PClass'] == '1st'].head(2)

Name PClass   Age     Sex  Survived  SexCode
0  Allen, Miss Elisabeth Walton    1st  29.0  female         1        1
1   Allison, Miss Helen Loraine    1st   2.0  female         0        1

<h1>2.3.1 Задание</h1>
<p>Найдитe евклидово расстояние между двумя Series (точками) a и b, не
используя встроенную формулу.</p>

In [ ]:
import pandas as pd
import numpy as np
print("Координаты точки a")
a1 = int(input("x = "))
b1 = int(input("y = "))
print("Координаты точки b\n")
a2 = int(input("x = "))
b2 = int(input("y = "))
ab = pd.Series([a1, b1, a2, b2])
print(np.sqrt((ab[0]-ab[2])**2 + (ab[1] - ab[3])**2))

Координаты точки a
x = 1
y = 1
Координаты точки b

x = 3
y = 3
2.8284271247461903


<h1> 2.3.2 Задание </h1>
<p>Найдите в Интернете ссылку на любой csv файл и сформируйте из него
фрейм данных (например, коллекцию фреймов данных можно найти
здесь: https://github.com/akmand/datasets).
</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/akmand/datasets/main/iris.csv'
dataframe = pd.read_csv(url)
dataframe

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm  \
0                5.1             3.5              1.4             0.2   
1                4.9             3.0              1.4             0.2   
2                4.7             3.2              1.3             0.2   
3                4.6             3.1              1.5             0.2   
4                5.0             3.6              1.4             0.2   
..               ...             ...              ...             ...   
145              6.7             3.0              5.2             2.3   
146              6.3             2.5              5.0             1.9   
147              6.5             3.0              5.2             2.0   
148              6.2             3.4              5.4             2.3   
149              5.9             3.0              5.1             1.8   

       species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]

<h1>2.3.3 Задание </h1>
<p>Проделайте с получившемся из предыдущего задания фреймом данных
те же действия, что и в примерах 2.2.5-2.2.7.</p>

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/akmand/datasets/main/iris.csv'
dataframe = pd.read_csv(url)
display(dataframe.head(5))

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm species
0              5.1             3.5              1.4             0.2  setosa
1              4.9             3.0              1.4             0.2  setosa
2              4.7             3.2              1.3             0.2  setosa
3              4.6             3.1              1.5             0.2  setosa
4              5.0             3.6              1.4             0.2  setosa

In [ ]:
display(dataframe.tail(3))


sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm  \
147              6.5             3.0              5.2             2.0   
148              6.2             3.4              5.4             2.3   
149              5.9             3.0              5.1             1.8   

       species  
147  virginica  
148  virginica  
149  virginica

In [ ]:
display(dataframe.shape)

(150, 5)

In [ ]:
display(dataframe.describe())

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm
count       150.000000      150.000000       150.000000      150.000000
mean          5.843333        3.054000         3.758667        1.198667
std           0.828066        0.433594         1.764420        0.763161
min           4.300000        2.000000         1.000000        0.100000
25%           5.100000        2.800000         1.600000        0.300000
50%           5.800000        3.000000         4.350000        1.300000
75%           6.400000        3.300000         5.100000        1.800000
max           7.900000        4.400000         6.900000        2.500000

In [ ]:
display(dataframe.iloc[1:4])

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm species
1              4.9             3.0              1.4             0.2  setosa
2              4.7             3.2              1.3             0.2  setosa
3              4.6             3.1              1.5             0.2  setosa

In [ ]:
dataframe[dataframe['sepal_length_cm'] == 5.1].head(3)

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm species
0               5.1             3.5              1.4             0.2  setosa
17              5.1             3.5              1.4             0.3  setosa
19              5.1             3.8              1.5             0.3  setosa

<h1>3.1. Теоретический материал – Работа с числовыми данными</h1>
<p>Количественные данные что-то измеряют – будь то размер класса,
ежемесячные продажи или оценки учащихся. Естественным способом
представления этих величин является численным (например, 150 студентов,
$529 392 продаж). </br>
Нормализация данных — это общепринятая задача предобработки в
машинном обучении. Многие алгоритмы предполагают, что все признаки
находятся в единой шкале, как правило, от 0 до 1 или от -1 до 1.</br>
Существует множество способов нормализации значений признаков,
чтобы масштабировать их к единому диапазону и использовать в различных
моделях машинного обучения. В зависимости от используемой функции, их
можно разделить на 2 большие группы: линейные и нелинейные. При
нелинейной нормализации в расчетных соотношениях используются
функции логистической сигмоиды или гиперболического тангенса. В
линейной нормализации изменение переменных осуществляется
пропорционально, по линейному закону.</br>
На практике наиболее распространены следующие методы
нормализации признаков:</br>
- Минимакс – линейное преобразование данных в диапазоне [0..1], где
минимальное и максимальное масштабируемые значения
соответствуют 0 и 1 соответственно;</br>
- Z-масштабирование данных на основе среднего значения и
стандартного отклонения: производят деление разницы между
переменной и средним значением на стандартное отклонение.</br>
При масштабировании данных мы будем использовать одну из
популярных библиотек машинного обучения Scikit-learn. Библиотека
содержит пакет sklearn.preprocessing, который предоставляет широкие
возможности для нормализации данных. Следует отметить, что в целом
алгоритмы обучения выигрывают от стандартизации набора данных.</p>

<h1> 3.2.1 Пример </h1>
<p>Прошкалируйте числовой признак в диапазон между двумя значениями.</p>

In [ ]:
import numpy as np
from sklearn import preprocessing
feature = np.array([[-10.5], [-90.2], [70], [90.1], [800.2]])
minmax_scale = preprocessing.MinMaxScaler(feature_range = (0, 1))
scaled_feature = minmax_scale.fit_transform(feature)
scaled_feature

array([[0.08951033],
       [0.        ],
       [0.17991914],
       [0.20249326],
       [1.        ]])

<h1>3.2.2 Пример </h1>
<p>Преобразуйте признак, чтобы он имел среднее значение 0 и стандартное
отклонение 1.</p>

In [ ]:
import numpy as np
from sklearn import preprocessing
x = np.array([[-1000.9], [-200.2], [500.5], [600.0], [9000.9]])
scaler = preprocessing.StandardScaler()
standardized = scaler.fit_transform(x)
print("Среднее: ", round(standardized.mean()))
print("Стандартное отклонение:", standardized.std())
standardized

Среднее:  0
Стандартное отклонение: 1.0


array([[-0.76069155],
       [-0.5416716 ],
       [-0.35000521],
       [-0.32278841],
       [ 1.97515676]])

<h1>3.2.3 Пример </h1>
<p>Дан фрейм данных </br>
dfTest = pd.DataFrame({'A':[14.00,90.20,90.95,96.27,91.21], </br>
 'B':[103.02,107.26,110.35,114.23,114.68], </br>
 'C':['big','small','big','small','small']}) </br>
Необходимо масштабировать его числовые столбцы.</p>

In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
dfTest = pd.DataFrame({ 'A': [14.00, 90.20, 90.95, 96.27, 91.21],
                        'B': [103.02, 107.26, 110.35, 114.23, 114.68],
                        'C': ['big', 'small', 'big', 'small', 'small']})
dfTest[['A', 'B']] = scaler.fit_transform(dfTest[['A', 'B']])
dfTest

A         B      C
0  0.000000  0.000000    big
1  0.926219  0.363636  small
2  0.935335  0.628645    big
3  1.000000  0.961407  small
4  0.938495  1.000000  small

<h1>3.3.2 Задание </h1>
<p>Загрузить фрейм данных по ссылке:
https://raw.githubusercontent.com/akmand/datasets/master/iris.csv.
Необходимо выполнить нормализацию первого числового признака
(sepal_length_cm) с использованием минимаксного преобразования, а
второго (sepal_width_cm) с задействованием z-масштабирования.</p>

In [ ]:
import pandas as pd
url ='https://raw.githubusercontent.com/akmand/datasets/master/iris.csv'
dataframe = pd.read_csv(url)
min_max = preprocessing.MinMaxScaler()
dataframe[["sepal_length_cm"]] = min_max.fit_transform(dataframe[["sepal_length_cm"]])
z = preprocessing.StandardScaler()
dataframe[["sepal_width_cm"]] = z.fit_transform(dataframe[["sepal_width_cm"]])
dataframe

sepal_length_cm  sepal_width_cm  petal_length_cm  petal_width_cm  \
0           0.222222        1.032057              1.4             0.2   
1           0.166667       -0.124958              1.4             0.2   
2           0.111111        0.337848              1.3             0.2   
3           0.083333        0.106445              1.5             0.2   
4           0.194444        1.263460              1.4             0.2   
..               ...             ...              ...             ...   
145         0.666667       -0.124958              5.2             2.3   
146         0.555556       -1.281972              5.0             1.9   
147         0.611111       -0.124958              5.2             2.0   
148         0.527778        0.800654              5.4             2.3   
149         0.444444       -0.124958              5.1             1.8   

       species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 5 columns]